cpu/nrf52/periph/i2c.c

```diff
@@ -0,0 +1,212 @@
+ *
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "assert.h"
+#include "periph/i2c.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG        (0)
+#include "debug.h"
+
+/**
+ * @brief   If any of the 4 lower bits are set, the speed value is invalid
+ */
+#define INVALID_SPEED_MASK  (0x0f)
+
+/**
+ * @brief   Initialized bus locks (we have a maximum of two devices...)
+ */
+static mutex_t locks[] = {
+    MUTEX_INIT,
+    MUTEX_INIT
+};
+
+static inline NRF_TWIM_Type *dev(i2c_t bus)
+{
+    return i2c_config[bus].dev;
+}
+
+static int error(i2c_t bus)
+{
+    DEBUG("[i2c] error 0x%02x\n", (int)dev(bus)->ERRORSRC);
+    dev(bus)->ERRORSRC = 0;
+    dev(bus)->EVENTS_ERROR = 0;
+    return -1;
+}
+
+static int write(i2c_t bus, uint8_t address, const void *data, int length, int stop)
+{
+    uint8_t *out_buf = (uint8_t *)data;
+    assert((bus <= I2C_NUMOF) && (length > 0));
+
+    DEBUG("[i2c]: writing %i byte to the bus\n", length);
+
+    /* disable shortcuts */
+    if (stop == 0) {
+        dev(bus)->SHORTS = 0;
+    }
+
+    /* set the client address and the data pointer */
+    dev(bus)->ADDRESS = (address & 0x7f);
+    dev(bus)->TXD.PTR = (uint32_t)out_buf;
+    dev(bus)->TXD.MAXCNT = (length << TWIM_TXD_MAXCNT_MAXCNT_Pos);
+
+    /* start write sequence */
+    dev(bus)->EVENTS_LASTTX = 0;
+    dev(bus)->TASKS_STARTTX = 1;
+
+    /* wait for the device to finish up */
+    while ((dev(bus)->EVENTS_LASTTX == 0) && (dev(bus)->EVENTS_ERROR == 0)) {}
+    if (dev(bus)->EVENTS_ERROR) {
+        return error(bus);
+    }
+
+    /* wait for the device to finish up */
+    while ((dev(bus)->TXD.AMOUNT != (unsigned)length) && (dev(bus)->EVENTS_ERROR == 0)) {}
+    if (dev(bus)->EVENTS_ERROR) {
+        return error(bus);
+    }
+
+    /* enable shortcuts */
+    if (stop == 0) {
+        dev(bus)->SHORTS = (TWIM_SHORTS_LASTTX_STOP_Enabled << TWIM_SHORTS_LASTTX_STOP_Pos) |
+                           (TWIM_SHORTS_LASTRX_STOP_Enabled << TWIM_SHORTS_LASTRX_STOP_Pos);
+    }
+
+    return dev(bus)->TXD.AMOUNT;
+}
+
+int i2c_init_master(i2c_t bus, i2c_speed_t speed)
+{
+    if (bus >= I2C_NUMOF) {
+        return -1;
+    }
+    if (speed & INVALID_SPEED_MASK) {
+        return -2;
+    }
+
+    /* pin configuration */
+    NRF_P0->PIN_CNF[i2c_config[bus].pin_scl] = (GPIO_PIN_CNF_DRIVE_S0D1 << GPIO_PIN_CNF_DRIVE_Pos);
+    NRF_P0->PIN_CNF[i2c_config[bus].pin_scl] = (GPIO_PIN_CNF_DRIVE_S0D1 << GPIO_PIN_CNF_DRIVE_Pos);
+    dev(bus)->PSEL.SCL = i2c_config[bus].pin_scl;
+    dev(bus)->PSEL.SDA = i2c_config[bus].pin_sda;
+
+    /* shortcuts configuration */
+    dev(bus)->SHORTS = (TWIM_SHORTS_LASTTX_STOP_Enabled << TWIM_SHORTS_LASTTX_STOP_Pos) |
+                       (TWIM_SHORTS_LASTRX_STOP_Enabled << TWIM_SHORTS_LASTRX_STOP_Pos);
+
+    /* bus clock speed configuration */
+    dev(bus)->FREQUENCY = (TWIM_FREQUENCY_FREQUENCY_K100 << TWIM_FREQUENCY_FREQUENCY_Pos);
```

Comment:
Please change `TWIM_FREQUENCY_FREQUENCY_K100` to `speed` so we can run I2C at 400 kHz.

Reply:
Good catch! I completely overlooked this one. Fixed in latest commit